django/apps/config.py

```diff
@@ -82,87 +87,157 @@ def _path_from_module(self, module):
     def create(cls, entry):
         """
         Factory that creates an app config from an entry in INSTALLED_APPS.
-        """
-        try:
-            # If import_module succeeds, entry is a path to an app module,
-            # which may specify an app config class with default_app_config.
-            # Otherwise, entry is a path to an app config class or an error.
-            module = import_module(entry)
-
-        except ImportError:
-            # Track that importing as an app module failed. If importing as an
-            # app config class fails too, we'll trigger the ImportError again.
-            module = None
-
-            mod_path, _, cls_name = entry.rpartition('.')
 
-            # Raise the original exception when entry cannot be a path to an
-            # app config class.
-            if not mod_path:
-                raise
+        """
+        # create() eventually returns app_config_class(app_name, app_module).
+        app_config_class = None
+        app_name = None
+        app_module = None
 
+        # If import_module succeeds, entry points to the app module.
+        try:
+            app_module = import_module(entry)
+        except Exception:
```

Comment:
Silencing far more exceptions than previously (where it was `ImportError`) ... is this intentional, and will it lead to people having a harder time debugging when their app config goes wrong? I guess it'd be somewhere higher up the chained stack traces, maybe?
Same expansion of caught exceptions (`ImportError` -> `Exception`) happens at the `import_string` so I assume the answer to this also answers that.

Reply:
All errors are ignored here (and a second time below).

Then there's logic to provide a good error message, whether the user tried the path to a module or an AppConfig class: https://github.com/django/django/pull/12310/files#diff-239ff7e40a4b9921b462471882a575f6R187-R209

Specifically, the same exception as here is re-raised there:

```python
                # Re-trigger the module import exception.
                import_module(entry)
```

Reply:
So, yes, this is intentional. It only depends on the assumption that running `import_module` twice doesn't make debugging significantly harder than running it once.

This assumption may be wrong in the presence e.g. of caching. If importing a Django application's module (not even the models module) or AppConfig class isn't idempotent, I feel like the user has been looking for trouble...

I think it's better to provide a nice error message for the common error case — typos — than to increase slightly the possibility that extremely rare and very disputable edge cases might be marginally easier to debug :-)